examples/.unstable/next-formdata/src/pages/api/trpc/[trpc].ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * This is the API-handler of your app that contains all your API routes.
+ * On a bigger app, you will probably want to split this file up into multiple files.
+ */
+import * as trpcNext from '@trpc/server/adapters/next';
+import {
+  nodeHTTPFormDataContentTypeHandler,
+  unstable_createMemoryUploadHandler,
+  unstable_parseMultipartFormData,
+} from '@trpc/server/adapters/node-http/content-type/form-data';
+import { nodeHTTPJSONContentTypeHandler } from '@trpc/server/adapters/node-http/content-type/json';
+import { NextApiRequest, NextApiResponse } from 'next';
+import fs from 'node:fs';
+import path from 'node:path';
+import { Readable } from 'node:stream';
+import { ReadableStream } from 'node:stream/web';
+import * as undici from 'undici';
+import { z } from 'zod';
+import { zfd } from 'zod-form-data';
+import { publicProcedure, router } from '~/server/trpc';
+import { uploadFileSchema } from '~/utils/schemas';
+
+globalThis.File = undici.File as any;
+
+async function writeFileToDisk(file: File) {
+  const rootDir = __dirname + '/../../../../..';
+
+  const nonce = Date.now();
+  const fileDir = path.resolve(`${rootDir}/public/uploads/${nonce}`);
+
+  if (!fs.existsSync(fileDir)) {
+    fs.mkdirSync(fileDir, { recursive: true });
+  }
+  console.log('Writing', file.name, 'to', fileDir);
+  const fd = fs.createWriteStream(path.resolve(`${fileDir}/${file.name}`));
+
+  for await (const chunk of Readable.fromWeb(file.stream() as ReadableStream)) {
+    fd.write(chunk);
+  }
+  fd.end();
+
+  return {
+    url: `/uploads/${nonce}/${file.name}`,
+    name: file.name,
+  };
+}
+
+function isPlainObject(obj: unknown): obj is object {
+  return !!obj && typeof obj === 'object' && !Array.isArray(obj);
+}
+
+const appRouter = router({
+  upload: publicProcedure
+    .use(async ({ ctx, next, input, rawInput }) => {
+      const formData = await unstable_parseMultipartFormData(
+        ctx.req,
+        unstable_createMemoryUploadHandler(),
+      );
+
+      return next({
+        rawInput: isPlainObject(rawInput)
+          ? {
+              ...rawInput,
+              ...Object.fromEntries(formData),
+            }
+          : formData,
+      } as any);
+    })
+    .input(
+      zfd.formData({
+        name: zfd.text(),
+        image: zfd.file(),
+        document: zfd.file(z.instanceof(File).optional()),
+      }),
+    )
+    .mutation(async (opts) => {
+      return {
+        image: await writeFileToDisk(opts.input.image),
```

Comment:
We might want it to be `opts.formData` so there's no colliding keys 🤔

Reply:
Scenario:

- You validate on `roomId=1`
- It's passed through a middleware that check you're in that room
- In `formData` you add a `roomId`
- The first `roomId` will be overwritten but not authed

Reply:
ah good point

Reply:
fixed

Reply:
I thought a bit more on this and it _is_ a bit ugly to have the `input.formData`

What do you think is nicest:

- have a `.formData` thing
- Spread `{...formData}` but throw an internal error if there's a colliding key existent? I think that'd be neater... WDYT?

Reply:
Given formdata is going to be the whole body if a mutation, I do think we're going to confuse users if we have to explain that sometimes your inputs are on `.input` and sometimes `.formData` and maybe sometimes `.stream`, shouldn't we endeavor to just adapt the inferred type of `input` in all cases?

When it comes to _returning_ blobs from functions that's also going to be necessary for `.output` right?